internal/cri/server/images/image_pull.go

```diff
@@ -798,3 +871,186 @@ func (c *CRIImageService) snapshotterFromPodSandboxConfig(ctx context.Context, i
 
 	return snapshotter, nil
 }
+
+type criCredentials struct {
+	ref         string
+	credentials func(string) (string, string, error)
+}
+
+func newcriCredentials(ref string, credentials func(string) (string, string, error)) registry.CredentialHelper {
+	return &criCredentials{
+		ref:         ref,
+		credentials: credentials,
+	}
+}
+
+// GetCredentials gets credential from criCredentials makes criCredentials a registry.CredentialHelper
+func (cc *criCredentials) GetCredentials(ctx context.Context, ref string, host string) (registry.Credentials, error) {
+	if cc.credentials == nil {
+		return registry.Credentials{}, fmt.Errorf("credential handler not initialized for ref %q", ref)
+	}
+
+	if ref != cc.ref {
+		return registry.Credentials{}, fmt.Errorf("invalid ref %q, expected %q", ref, cc.ref)
+	}
+
+	username, secret, err := cc.credentials(host)
+	if err != nil {
+		return registry.Credentials{}, fmt.Errorf("failed to get credentials for %q: %w", host, err)
+	}
+	return registry.Credentials{
+		Host:     host,
+		Username: username,
+		Secret:   secret,
+	}, nil
+}
+
+type transferProgressReporter struct {
+	ref               string
+	pc                chan transfer.Progress
+	cancel            context.CancelFunc
+	timeout           time.Duration
+	reqReporter       pullRequestReporter
+	statuses          map[string]*transfer.Progress
+	lastSeenBytesRead uint64
+	lastSeenTimestamp time.Time
+}
```

Comment:
Is this something we can reuse with the transfer service pull in ctr? or cri specific logic?

Reply:
This is CRI specific. `ctr` actually has its own progress reporter which has more detailed display for layer image pull progress. CRI progress reporter on the other hand is quite basic and is mainly used to track image progress timeouts. I have kept the functionality similar to how the current progress is reported for CRI.